app/frontend/src/cljs/teet/ui/query.cljs

```diff
@@ -46,38 +49,50 @@
         state-atom (when-not state-path
                      (r/atom nil))
+        filter-atom (r/atom {})
+        debounce (functions/debounce (fn []
+                                       (e! (common-controller/->Refresh)))
+                                     250)
+        change-filter #(swap! filter-atom merge %)
+        reset-filter #(reset! filter-atom {})
         poll-id (when poll-seconds
                   (js/setInterval #(e! (->Query query args state-path state-atom))
                                   (* 1000 poll-seconds)))]
+    (add-watch filter-atom :requery-filters
+               (fn [_ _ _ _]
+                 (debounce)))
     (e! (->Query query args state-path state-atom))
     (r/create-class
-     {:component-will-unmount #(do
-                                 (when poll-id
-                                   (js/clearInterval poll-id))
-                                 (e! (->Cleanup state-path)))
-      :reagent-render
-      (fn [{:keys [e! query args state-path skeleton view app state refresh breadcrumbs
-                   simple-view loading-state]}]
-        (let [state (if state-path
-                      state
-                      @state-atom)]
-          (when (or (not= @refresh-value refresh)
-                    (not= @previous-args args))
-            (reset! refresh-value refresh)
-            (reset! previous-args args)
-            (e! (->Query query args state-path state-atom)))
-          (if (or state loading-state)
-            ;; Results loaded, call the view
-            (if simple-view
-              (conj simple-view (or state loading-state))
-              ^{:key "query-result-view"}
-              [query-page-view view e! app state breadcrumbs])
+      {:component-will-unmount #(do
+                                  (when poll-id
+                                    (js/clearInterval poll-id))
+                                  (e! (->Cleanup state-path)))
+       :reagent-render
+       (fn [{:keys [e! query args state-path skeleton view app state refresh breadcrumbs
+                    simple-view loading-state]}]
+         [context/provide :query-filter {:value @filter-atom
+                                         :on-change change-filter
+                                         :reset-filter reset-filter}
+          (let [state (if state-path
+                        state
+                        @state-atom)]
+            (when (or (not= @refresh-value refresh)
+                      (not= @previous-args args))
+              (reset! refresh-value refresh)
+              (reset! previous-args args)
+              (e! (->Query query (assoc args :filters @filter-atom) state-path state-atom)))
+            (if (or state loading-state)
+              ;; Results loaded, call the view
+              (if simple-view
+                (conj simple-view (or state loading-state))
+                ^{:key "query-result-view"}
+                [query-page-view view e! app state breadcrumbs])
```

Comment:
I would refactored this into _render_ separated function to make _query_ function easier to read.